### PR TITLE
Wd 286  optimize installer

### PR DESCRIPTION
**Pull request by nithinsaii**
- Features added:
1. un-installation of an existing instance of agent during installation
2. made it interactive

- docker installation script added

### CODE REVIEW

File: weeve-agent-installer.sh [hunk 1]
```diff
@@ -22,91 +25,114 @@ cleanup() {
...
 
 # downloading the dependencies with personal access key since its stored in private repository
-for dependency in AmazonRootCA1.pem 4be43aa6f1-certificate.pem.crt 4be43aa6f1-private.pem.key nodeconfig.json weeve-agent.service weeve-agent.argconf
+for DEPENDENCIES in AmazonRootCA1.pem 1d77ae9afd-certificate.pem.crt 1d77ae9afd-private.pem.key nodeconfig.json weeve-agent.service weeve-agent.argconf
 do
-if result=$(cd ./weeve-agent \
-&& curl -sO https://"$access_key"@raw.githubusercontent.com/weeveiot/weeve-agent-dependencies/master/$dependency 2>&1); then
-  log $dependency downloaded.
+if RESULT=$(cd ./weeve-agent \
+&& curl -sO https://"$ACCESS_KEY"@raw.githubusercontent.com/nithinsaii/weeve-agent-dependencies--demo/master/$DEPENDENCIES 2>&1); then
```
Comment (shriftz) [inline weeve-agent-installer.sh]: Shall we keep it weeveiot instead of person name (nithinsaii)

File: weeve-agent-installer.sh [hunk 2]
```diff
@@ -22,91 +25,114 @@ cleanup() {
...
 
 # checking for the file containing access key
-if [ -f "$secret_file" ];then
+if [ -f "$SECRET_FILE" ];then
 log Reading the access key ...
-access_key=$(cat "$secret_file")
+ACCESS_KEY=$(cat "$SECRET_FILE")
 else
-log Please create and file named '.weeve-agent-secret' and append the access key of the github into the file!!!
+log File not found!!!
```
Comment (shriftz) [inline weeve-agent-installer.sh]: Please make it more specific

File: weeve-agent-installer.sh [hunk 3]
```diff
@@ -22,91 +25,114 @@ cleanup() {
...
+ARGUMENTS_FILE=/lib/systemd/system/weeve-agent.argconf
 
-secret_file=.weeve-agent-secret
+if [ -d "$WEEVE_AGENT_DIRECTORY" ] || [ -f "$SERVICE_FILE" ] || [ -f "$ARGUMENTS_FILE" ]; then
+  log Detected some weeve-agent contents!
+  log Proceeding with the un-installation of the existing instance of weeve-agent ...
+  cleanup
+  log Continuing with the installation ...
+else
+  log No weeve-agent contents found, proceeding with the installation ...
```
Comment (shriftz) [inline weeve-agent-installer.sh]: Do we need this part?

File: weeve-agent-installer.sh [hunk 4]
```diff
@@ -22,91 +25,114 @@ cleanup() {
...
+if [ -z "$NODE_NAME" ]; then
+log NODE_NAME is required
+read -p "Give a node name: " NODE_NAME
 fi
 
 log All arguments are set
-log Name of the node: "$node_name"
+log Name of the node: "$NODE_NAME"
+
+log Checking if a instance of weeve-agent is already running ...
```
Comment (shriftz) [inline weeve-agent-installer.sh]: I think we don't need this log for user

File: weeve-agent-installer.sh [hunk 5]
```diff
@@ -22,91 +25,114 @@ cleanup() {
...
+
+CURRENT_DIRECTORY=$(pwd)
+
+WEEVE_AGENT_DIRECTORY="$CURRENT_DIRECTORY"/weeve-agent
+SERVICE_FILE=/lib/systemd/system/weeve-agent.service
+ARGUMENTS_FILE=/lib/systemd/system/weeve-agent.argconf
 
-secret_file=.weeve-agent-secret
+if [ -d "$WEEVE_AGENT_DIRECTORY" ] || [ -f "$SERVICE_FILE" ] || [ -f "$ARGUMENTS_FILE" ]; then
+  log Detected some weeve-agent contents!
```
Comment (shriftz) [inline weeve-agent-installer.sh:55]: It's better to let user decide, prompt user message that weeveAgent is already installed and ask if they want to continue or cancel installation (providing some options), just like we see while installing any application in Linux using command prompt.
Follow the same at all places as needed

File: weeve-agent-installer.sh [hunk 6]
```diff
@@ -115,48 +145,47 @@ log Dependencies downloaded.
...
 log Starting the service ...
 
 # moving .service and .argconf to systemd path and starting the service
-if result=$(sudo mv weeve-agent/weeve-agent.service /lib/systemd/system/ \
+if RESULT=$(sudo mv weeve-agent/weeve-agent.service /lib/systemd/system/ \
 && sudo mv weeve-agent/weeve-agent.argconf /lib/systemd/system/ \
 && sudo systemctl enable weeve-agent \
 && sudo systemctl start weeve-agent 2>&1); then
-  log weeve-agent service should be up, you will be prompted once weeve-agent is connected.
+  log weeve-agent service should be up in around 3 MINUTES, you will be prompted once weeve-agent is connected.
```
Comment (shriftz) [inline weeve-agent-installer.sh]: can you reduce it to 30-60 sec

File: weeve-agent-installer.sh [hunk 7]
```diff
@@ -115,48 +145,47 @@ log Dependencies downloaded.
 
 # appending the argument for node name to weeve-agent.argconf
 log Adding the node name argument ...
-echo "ARG_NODENAME=--name $node_name" >> ./weeve-agent/weeve-agent.argconf
+echo "ARG_NODENAME=--name $NODE_NAME" >> ./weeve-agent/weeve-agent.argconf
 
 # appending the required strings to the .service to point systemd to the path of the binary
 # following are the lines appended to weeve-agent.service
 # WorkingDirectory=/home/nithin/weeve-agent
 # ExecStart=/home/nithin/weeve-agent/weeve-agent-x86_64 $ARG_VERBOSE $ARG_BROKER $ARG_SUB_CLIENT $ARG_PUB_CLIENT $ARG_PUBLISH $ARG_HEARTBEAT $ARG_NODENAME
```
Comment (shriftz) [inline weeve-agent-installer.sh]: Please don't use specific names

**Review by ersanyamarya — APPROVED** [review]
Looks good to me

File: weeve-agent-installer.sh (hunk 6 above)
Comment (nithinsaii) [reply weeve-agent-installer.sh]: will be handled in agent

**Pull request merged**